### DISCUSSION TITLE
Adding a compiled/released package?

### DISCUSSION THREAD

**Discussion by SoylentGraham** [openupm/openupm#1872] [Q&A]
I have an open source package, which is compiled (c++ for win/mac/magicleap/android/linux/ios etc)

This builds to a release, (when I commit a `vX.Y.Z` tag) and an npm package on github's npm registry. I currently include this package the same way openupm works, but as we know, there are big user painpoints adding a scoped registry in unity.
Until this is fixed in unity, I was wondering if it would be worth distributing via OpenUpm too (hoping a user has already setup openupm once)

But from what I can tell, you only really support code that would be committed to a branch.

Have I misunderstood? Or is it possible to distribute my project via openupm?
https://github.com/newchromantics/poph264

**Comment by favoyang** [comment]
Hi @SoylentGraham,

I think the best way to go is to prepare a ready-to-use package branch, like `upm` which included all complied DLLs using GitHub Actions.
- You can only do that when pushing a versioned tag `x.y.z`, then build tag `upm/x.y.z`.
- You can optional to build a `upm` branch synced with the main branch. Then you get a bloated repository to keep the binary file's history. You can use a strategy to delete the old upm branch, recreate it from the scratch, then commit with `-f` to keep a single commit history only.

For OpenUPM, it supports a Git tag prefix to only track certain tags, like `upm/`.

I prefer making the OpenUPM build-pipelines as simple as a packing service. Because the custom build process is usually too flexible. You gonna have some trial-and-error processes to make it work. Having such processes on OpenUPM pipelines can be suffering.

**Reply by SoylentGraham** [reply]
Oh, that's not a practical solution to me I'm afraid.
- I believe having different content on different branches is bad practise.
- Manually cleaning that branch to avoid running out of space... is messy and dangerous when automating it :)

I guess my only practical solution is to have a seperate repostiory to commit binaries to... but at that point, I don't need OpenUPM :) (as I can just use that git url in unity directly and bypassing npm/registry)

Ah well! maybe in the future

**Reply by favoyang** [reply]
I understand your concern about some of the "creative" usages of Git. Well, basically anything on the OpenUPM can be replaced by the Git URL with #tag-name. Just being a registry package has some benefits over Git URL like the better Packman integration.

It just the custom build script is too flexible. It can be a windows script, a shell script that only runs on Mac OS, or whatever.  I don't want to invent the whole thing just for this project. Using GitHub Actions is a better choice.

I guess one way is that let OpenUPM support the traditional registry's "push-mode" where the user creates an account and publishes a package to the registry via `npm publish`, like any trivial registry server. But not at this point.

**Comment by SoylentGraham** [comment]
Im not sure if theres a misunderstanding here, Im not asking openupm to
build my binaries, nor any custom script.
My repository already builds everything via github actions. (To github
releases, GitHubs npm registry, soon to the asset store, etc)

What I was hoping was that openupm could fetch the assets via say, http
fetch of the url to the npm-package tgz in the github registery, instead of
fetching from a git repository.
Obviously its not exactly the same, but I'd imagine its just the "fetch
files" part of the pipeline. (Of course I don't know how you've structured
it right now)

The only real benefit this integration would be, would be so users only
need to setup openupm once, rather than another registry over and over
(still a big ux oversight by unity i think :)

On Fri, 5 Feb 2021 at 12:58 pm, Favo Yang <notifications@github.com> wrote:

> I understand your concern about some of the "creative" usages of Git.
> Well, basically anything on the OpenUPM can be replaced by the Git URL with
> #tag-name. Just being a registry package has some benefits over Git URL
> like the better Packman integration.
>
> It just the custom build script is too flexible. It can be a windows
> script, a shell script that only runs on Mac OS, or whatever. I don't want
> to invent the whole thing just for this project. Using GitHub Actions is a
> better choice.
>
> I guess one way is that let OpenUPM support the traditional registry's
> "push-mode" where the user creates an account and publishes a package to
> the registry via npm publish, like any trivial registry server. But not
> at this point.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openupm/openupm/discussions/1872#discussioncomment-341488>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQVIBJPSPM2ASGRJASUMVTS5PTOHANCNFSM4XDHPEWQ>
> .
>
-- 

- Graham Reeves
panopo.ly <https://www.panopo.ly/>

**Reply by favoyang** [reply]
> What I was hoping was that openupm could fetch the assets via say, http fetch of the url to the npm-package tgz in the github registery, instead of fetching from a git repository. Obviously its not exactly the same, but I'd imagine its just the "fetch files" part of the pipeline. (Of course I don't know how you've structured it right now)

Not sure if I understand correctly
- "fetch some files (like DLLs), place them into some paths, then create the tarball to publish to OpenUPM" is already a custom build process. We need to create a language/configuration for that. That's what I'm trying to avoid - a build language/configuration that only works for OpenUPM. We even don't have a way to tell the pipelines to move the `Example` folder to `Example~` folder. Instead, we suggest the "packaging-only" strategy described in my first reply https://github.com/openupm/openupm/discussions/1872#discussioncomment-340803. I'm not entirely rejecting the openupm configuration file idea, just be more cautious.
- If you mean fetch the built tarball directly from another registry like the GitHub registry. That turns OpenUPM to be a proxy registry. An interesting idea, but it could be abused. We're in the process to apply Unity Package TOS to the entire website, anything that may break it, has to be delayed. https://openupm.com/docs/adding-upm-package.html

BTW, the build-pipelines are open-source at https://github.com/openupm/openupm-pipelines/blob/master/azure-pipelines.yml

**Reply by atteneder** [reply]
How about fetching the GH Action built tarball from release artifacts then?

Alternatively I liked your idea of pushing to OpenUPM via `npm publish` in a GH action that's triggered on releases (maybe even better).

I suffer the same problems and would love a better solution here. Since my binaries are significant in size, I commit them via GIT LFS. GitHub charges me heavily for that traffic. This could be avoided completely with proper build job support.

Thanks for considering.

**Reply by atteneder** [reply]
In other words, I'd rather raise my monthly Patreon donations to you instead of throwing money at GitHub/Microsoft :)

**Reply by favoyang** [reply]
After revisiting the thread, we have three options.

Op1: Fetch artifacts from a GitHub release.
- Add the `trackingMode` boolean property to manage the tracking source: github or githubRelease.
- Add the `GitHubArtifactFilename` property if a release contains more than one artifact.
- The pipeline will detect a new Git tag, then wait for the first GitHub release entry based on the Git tag to fetch the artifact.

Op2:  Enable the registry push mode.
- It turns the registry into a traditional way to let users run `npm publish` directly.
- It introduces more spam content, as people will randomly try it out.
- It requires a trivial user account system that supports registering, signing in, signing out, token management, forgetting passwords, 2FA...
- Need to find a way to avoid conflicts with the existing pull mode packages.
- Need to upgrade our infrastructure. however, I'm not optimistic about the budget.

Op3: Sync with a custom uplink registry (allowing you to publish to the GitHub registry or the npm registry and then sync back to openupm)
- add the `trackingMode` boolean property to manage the tracking source: github or uplink.
- add the `uplinkRegistryBaseUrl` property, e.g. https://registry.npmjs.org.
- the pipeline will monitor the uplink registry package regularly and then sync back to the OpenUPM registry.

I prefer Op1, considering it is the most feasible option that is compatible with the way OpenUPM is operating.

I will create a ticket to track the feature. Though no ETA.

**Reply by favoyang** [reply]
Tracked in https://github.com/openupm/openupm/issues/3923

**Reply by SoylentGraham** [reply]
Sorry for going quiet on this topic a while ago.

FYI, for an example of a package system that does fetch binaries from another external source, as well as having source etc; Swift package manager does this for xcode (and works very well, I wish all package managers had something similar)
An example here; https://github.com/NewChromantics/PopEngine/blob/master/Package.swift (obviously apple's docs have more info)

**Reply by favoyang** [reply]
For npm, prebuilt binaries are used to implement as a hook on the `scripts` entry's `install` command:

```
{
  "scripts": {
    "install": "prebuild-install || node-gyp rebuild"
  }
}
```

Using the tool `prebuild` to upload binaries to GitHub, and using `prebuild-install` to retrieve it, or fallback to `node-gyp rebuild`.

Another combination is [prebuildify](https://github.com/prebuild/prebuildify ) and [node-gyp-build](https://github.com/prebuild/node-gyp-build), which stores binaries for multiple platforms within a single npm package itself. It will generate a fat package usually suitable for a small codebase/module.

Unity reuses the npm toolchain, but it doesn't support the package hook lifecycle, and it also lacks an official way to fetch binaries for different platforms. All you can do is handle it yourself, either prebuilt all platform-specific binaries and ship with the package itself or create custom scripts (even some UI popups) to handle post-install for binaries (by downloading them somewhere in the Assets folder).

**Comment by PereViader** [comment]
I would also find this very useful.
https://github.com/openupm/openupm/discussions/4918
Needing to have a UPM branch / another repository for this is really not ideal.
